Content.Shared/_CM14/Scoping/SharedScopeSystem.cs

```diff
@@ -0,0 +1,193 @@
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        InitializeUser();
+        SubscribeLocalEvent<ScopeComponent, GotEquippedHandEvent>(OnEquip);
+        SubscribeLocalEvent<ScopeComponent, GotUnequippedHandEvent>(OnUnequip);
+        SubscribeLocalEvent<ScopeComponent, DroppedEvent>(OnDrop);
+        SubscribeLocalEvent<ScopeComponent, HandDeselectedEvent>(OnDeselectHand);
+
+        SubscribeLocalEvent<ScopeComponent, GetItemActionsEvent>(OnGetActions);
+        SubscribeLocalEvent<ScopeComponent, ToggleActionEvent>(OnToggleAction);
+
+        SubscribeLocalEvent<ScopeComponent, ComponentShutdown>(OnShutdown);
+
+        SubscribeLocalEvent<ScopeComponent, MapInitEvent>(OnMapInit);
+    }
+
+    private void OnMapInit(EntityUid uid, ScopeComponent component, MapInitEvent args)
+    {
+        _actionContainer.EnsureAction(uid, ref component.ScopingToggleActionEntity, component.ScopingToggleAction);
+        Dirty(uid, component);
+    }
+
+    private void OnEquip(EntityUid uid, ScopeComponent component, GotEquippedHandEvent args)
+    {
+        component.User = args.User;
+        Dirty(uid, component);
+
+        if(_net.IsServer)
+            EnsureComp<ScopeUserComponent>(args.User);
+    }
+
+    private void OnUnequip(EntityUid uid, ScopeComponent component, GotUnequippedHandEvent args)
+    {
+        StopScopingHelper(uid, component, args.User);
+    }
+
+    private void OnDrop(EntityUid uid, ScopeComponent component, DroppedEvent args)
+    {
+        StopScopingHelper(uid, component, args.User);
+    }
+
+    private void OnDeselectHand(EntityUid uid, ScopeComponent component, HandDeselectedEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        if (component.IsScoping)
+            StopScoping(uid, component, args.User);
+    }
+
+    private void OnGetActions(EntityUid uid, ScopeComponent component, GetItemActionsEvent args)
+    {
+        args.AddAction(ref component.ScopingToggleActionEntity, component.ScopingToggleAction);
+    }
+
+    private void OnToggleAction(EntityUid uid, ScopeComponent component, ToggleActionEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        if (!(_handsSystem.TryGetActiveItem(args.Performer, out var heldItem) && heldItem == uid))
+        {
+            var msgError = Loc.GetString("cm-action-popup-scoping-user-must-hold", ("scope", Name(uid)));
+            if (_gameTiming.IsFirstTimePredicted && _gameTiming.InPrediction)
+            {
+                _popupSystem.PopupEntity(msgError, args.Performer, args.Performer);
+            }
+
+            return;
+        }
+
+        if (component.IsScoping)
+            StopScoping(uid, component, args.Performer);
+        else
+            StartScoping(uid, component, args.Performer);
+
+        args.Handled = true;
+    }
+
+    private void OnShutdown(EntityUid uid, ScopeComponent component, ComponentShutdown args)
```

Comment:
For all of these event handlers you can replace `EntityUid uid, ScopeComponent component, ComponentShutdown args` with `Entity<ScopeComponent> ent, ref ComponentShutdown args` which is the preferred way to handle events.
It couples the component to the entity uid so it doesn't get accidentally passed around with the wrong entity uid and component combination, and it makes you not have to remember if an event is handled by ref or not. See all usages of SubscribeLocalEvent for ones that you can convert.

Reply:
Fixed, thanks!